### PR TITLE
[Backport][Pipelines] Noop Writes when Reader completed

### PR DESCRIPTION
**Pull request by BrennanConroy**
Backport of https://github.com/dotnet/runtime/pull/41071

## Customer Impact
Customers shouldn't be able to see any differences as we want to keep existing code working without modifications.
This is fixing a Pipelines usage pattern that could result in large memory pressure or OOMs in the worst case.

## Regression
No.

## Testing
Yes? Tests were written and existing tests required no changes.

## Risk
Low, change was purposefully tailored to not break existing code

### CODE REVIEW

**Review by davidfowl — APPROVED** [review]

**Comment by BrennanConroy** [comment]
@Anipik Could you take a look to make sure I added the change to the build correctly?

File: src/packages.builds [hunk 1]
```diff
@@ -26,6 +26,9 @@
       <AdditionalProperties>$(AdditionalProperties)</AdditionalProperties>
     </Project>
     <!-- add specific builds / pkgproj's here to include in servicing builds -->
+      <Project Include="$(MSBuildThisFileDirectory)System.IO.Pipelines\pkg\System.IO.Pipelines.pkgproj">
```
Comment (Anipik) [inline src/packages.builds]: nit: indentation

**Review by Anipik — CHANGES_REQUESTED** [review]
You need to update the package version and assembly version here https://github.com/dotnet/corefx/blob/release/3.1/src/System.IO.Pipelines/Directory.Build.props#L4-L5

File: src/System.IO.Pipelines/Directory.Build.props [hunk 2]
```diff
@@ -1,8 +1,8 @@
 ﻿<Project>
   <Import Project="..\Directory.Build.props" />
   <PropertyGroup>
-    <AssemblyVersion>4.0.2.1</AssemblyVersion>
-    <PackageVersion>4.7.2</PackageVersion>
+    <AssemblyVersion>4.0.2.2</AssemblyVersion>
+    <PackageVersion>4.7.3</PackageVersion>
```
Comment (BrennanConroy) [inline src/System.IO.Pipelines/Directory.Build.props:5]: Don't understand these versions at all

Comment (Anipik) [reply src/System.IO.Pipelines/Directory.Build.props:5]: https://github.com/dotnet/corefx/blob/release/3.1/Documentation/project-docs/library-servicing.md
can give u more context.

File: pkg/Microsoft.Private.PackageBaseline/packageIndex.json [hunk 3]
```diff
@@ -2583,7 +2583,8 @@
         "4.5.3",
         "4.6.0",
         "4.7.0",
-        "4.7.1"
+        "4.7.1",
+        "4.7.2"
```
Comment (Anipik) [inline pkg/Microsoft.Private.PackageBaseline/packageIndex.json]: nit: add "4.7.3"

File: pkg/Microsoft.Private.PackageBaseline/packageIndex.json [hunk 4]
```diff
@@ -2592,7 +2593,8 @@
         "4.0.0.1": "4.5.2",
         "4.0.1.0": "4.6.0",
         "4.0.2.0": "4.7.0",
-        "4.0.2.1": "4.7.1"
+        "4.0.2.1": "4.7.1",
+        "4.0.2.2": "4.7.2"
```
Comment (Anipik) [inline pkg/Microsoft.Private.PackageBaseline/packageIndex.json]: this should "4.7.3"

**Review by Anipik — APPROVED** [review]
left 2 comments

**Comment by Anipik** [comment]
They have been approved offline

**Pull request merged**